File: tests/sprokit/pipeline/test_scheduler_registry.cxx
/*ckwg +5
 * Copyright 2011-2013 by Kitware, Inc. All Rights Reserved. Please refer to
 * KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
 * Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.
 */

#include <test_common.h>

#include <sprokit/pipeline/config.h>
#include <sprokit/pipeline/modules.h>
#include <sprokit/pipeline/pipeline.h>
#include <sprokit/pipeline/scheduler.h>
#include <sprokit/pipeline/scheduler_registry.h>
#include <sprokit/pipeline/scheduler_registry_exception.h>
#include <sprokit/pipeline/types.h>

#include <boost/foreach.hpp>
#include <boost/make_shared.hpp>

#define TEST_ARGS ()

DECLARE_TEST(get_twice);
DECLARE_TEST(null_config);
DECLARE_TEST(null_pipeline);
DECLARE_TEST(load_schedulers);
DECLARE_TEST(null_ctor);
DECLARE_TEST(duplicate_types);
DECLARE_TEST(unknown_types);
DECLARE_TEST(module_marking);

int
main(int argc, char* argv[])
{
  CHECK_ARGS(1);

  testname_t const testname = argv[1];

  DECLARE_TEST_MAP(tests);

  ADD_TEST(tests, get_twice);
  ADD_TEST(tests, null_config);
  ADD_TEST(tests, null_pipeline);
  ADD_TEST(tests, load_schedulers);
  ADD_TEST(tests, null_ctor);
  ADD_TEST(tests, duplicate_types);
  ADD_TEST(tests, unknown_types);
  ADD_TEST(tests, module_marking);

  RUN_TEST(tests, testname);
}

IMPLEMENT_TEST(get_twice)
{
  sprokit::scheduler_registry_t const reg1 = sprokit::scheduler_registry::self();
  sprokit::scheduler_registry_t const reg2 = sprokit::scheduler_registry::self();

  if (reg1 != reg2)
  {
    TEST_ERROR("Received two different registries");
  }
}

IMPLEMENT_TEST(null_config)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::config_t const config;
  sprokit::pipeline_t const pipe;

  EXPECT_EXCEPTION(sprokit::null_scheduler_registry_config_exception,
                   reg->create_scheduler(sprokit::scheduler_registry::type_t(), pipe, config),
                   "requesting a NULL config to a scheduler");
}

IMPLEMENT_TEST(null_pipeline)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::config_t const config = sprokit::config::empty_config();
  sprokit::pipeline_t const pipe;

  EXPECT_EXCEPTION(sprokit::null_scheduler_registry_pipeline_exception,
                   reg->create_scheduler(sprokit::scheduler_registry::type_t(), pipe),
                   "requesting a NULL pipeline to a scheduler with default arguments");

  EXPECT_EXCEPTION(sprokit::null_scheduler_registry_pipeline_exception,
                   reg->create_scheduler(sprokit::scheduler_registry::type_t(), pipe, config),
                   "requesting a NULL pipeline to a scheduler");
}

IMPLEMENT_TEST(load_schedulers)
{
  sprokit::load_known_modules();

  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::scheduler_registry::types_t const types = reg->types();

  sprokit::pipeline_t const pipe = boost::make_shared<sprokit::pipeline>();

  BOOST_FOREACH (sprokit::scheduler_registry::type_t const& type, types)
  {
    sprokit::scheduler_t scheduler;

    try
    {
      scheduler = reg->create_scheduler(type, pipe);
    }
    catch (sprokit::no_such_scheduler_type_exception const& e)
    {
      TEST_ERROR("Failed to create scheduler: " << e.what());

      continue;
    }
    catch (std::exception const& e)
    {
      TEST_ERROR("Unexpected exception when creating scheduler: " << e.what());

      continue;
    }

    if (!scheduler)
    {
      TEST_ERROR("Received NULL scheduler (" << type << ")");

      continue;
    }

    if (reg->description(type).empty())
    {
      TEST_ERROR("The description for "
                 << type << " is empty");
    }
  }
}

IMPLEMENT_TEST(null_ctor)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  EXPECT_EXCEPTION(sprokit::null_scheduler_ctor_exception,
                   reg->register_scheduler(sprokit::scheduler_registry::type_t(), sprokit::scheduler_registry::description_t(), sprokit::scheduler_ctor_t()),
                   "requesting an non-existent scheduler type");
}

static sprokit::scheduler_t null_scheduler(sprokit::pipeline_t const& pipeline, sprokit::config_t const& config);

IMPLEMENT_TEST(duplicate_types)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::scheduler_registry::type_t const non_existent_scheduler = sprokit::scheduler_registry::type_t("no_such_scheduler");

  reg->register_scheduler(non_existent_scheduler, sprokit::scheduler_registry::description_t(), null_scheduler);

  EXPECT_EXCEPTION(sprokit::scheduler_type_already_exists_exception,
                   reg->register_scheduler(non_existent_scheduler, sprokit::scheduler_registry::description_t(), null_scheduler),
                   "requesting an non-existent scheduler type");
}

IMPLEMENT_TEST(unknown_types)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::scheduler_registry::type_t const non_existent_scheduler = sprokit::scheduler_registry::type_t("no_such_scheduler");

  sprokit::pipeline_t const pipe = boost::make_shared<sprokit::pipeline>();

  EXPECT_EXCEPTION(sprokit::no_such_scheduler_type_exception,
                   reg->create_scheduler(non_existent_scheduler, pipe),
                   "requesting an non-existent scheduler type");

  EXPECT_EXCEPTION(sprokit::no_such_scheduler_type_exception,
                   reg->description(non_existent_scheduler),
                   "requesting an non-existent scheduler type");
}

IMPLEMENT_TEST(module_marking)
{
  sprokit::scheduler_registry_t const reg = sprokit::scheduler_registry::self();

  sprokit::scheduler_registry::module_t const module = sprokit::scheduler_registry::module_t("module");

  if (reg->is_module_loaded(module))
  {
    TEST_ERROR("The module \'" << module << "\' is "
               "already marked as loaded");
  }

  reg->mark_module_as_loaded(module);

  if (!reg->is_module_loaded(module))
  {
    TEST_ERROR("The module \'" << module << "\' is "
               "not marked as loaded");
  }
}

sprokit::scheduler_t
null_scheduler(sprokit::pipeline_t const& /*pipeline*/, sprokit::config_t const& /*config*/)
{
  return sprokit::scheduler_t();
}

File: tests/bindings/python/sprokit/pipeline/test-scheduler_registry.py
#!@PYTHON_EXECUTABLE@
#ckwg +4
# Copyright 2011-2013 by Kitware, Inc. All Rights Reserved. Please refer to
# KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
# Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.


def test_import():
    try:
        from sprokit.pipeline import config
        import sprokit.pipeline.scheduler_registry
    except:
        test_error("Failed to import the scheduler_registry module")


def test_create():
    from sprokit.pipeline import config
    from sprokit.pipeline import scheduler_registry

    scheduler_registry.SchedulerRegistry.self()
    scheduler_registry.SchedulerType()
    scheduler_registry.SchedulerTypes()
    scheduler_registry.SchedulerDescription()
    scheduler_registry.SchedulerModule()


def test_api_calls():
    from sprokit.pipeline import config
    from sprokit.pipeline import modules
    from sprokit.pipeline import pipeline
    from sprokit.pipeline import scheduler_registry

    modules.load_known_modules()

    reg = scheduler_registry.SchedulerRegistry.self()

    sched_type = 'thread_per_process'
    c = config.empty_config()
    p = pipeline.Pipeline()

    reg.create_scheduler(sched_type, p)
    reg.create_scheduler(sched_type, p, c)
    reg.types()
    reg.description(sched_type)
    reg.default_type


def example_scheduler(check_init):
    from sprokit.pipeline import scheduler

    class PythonExample(scheduler.PythonScheduler):
        def __init__(self, pipe, conf):
            scheduler.PythonScheduler.__init__(self, pipe, conf)

            self.ran_start = check_init
            self.ran_wait = check_init
            self.ran_stop = check_init
            self.ran_pause = check_init
            self.ran_resume = check_init

        def _start(self):
            self.ran_start = True

        def _wait(self):
            self.ran_wait = True

        def _stop(self):
            self.ran_stop = True

        def _pause(self):
            self.ran_pause = True

        def _resume(self):
            self.ran_resume = True

        def __del__(self):
            if not self.ran_start:
                test_error("start override was not called")
            if not self.ran_wait:
                test_error("wait override was not called")
            if not self.ran_stop:
                test_error("stop override was not called")
            if not self.ran_pause:
                test_error("pause override was not called")
            if not self.ran_resume:
                test_error("resume override was not called")

    return PythonExample


def test_register():
    from sprokit.pipeline import config
    from sprokit.pipeline import modules
    from sprokit.pipeline import pipeline
    from sprokit.pipeline import scheduler_registry

    modules.load_known_modules()

    reg = scheduler_registry.SchedulerRegistry.self()

    sched_type = 'python_example'
    sched_desc = 'simple description'

    reg.register_scheduler(sched_type, sched_desc, example_scheduler(True))

    if not sched_desc == reg.description(sched_type):
        test_error("Description was not preserved when registering")

    p = pipeline.Pipeline()

    try:
        s = reg.create_scheduler(sched_type, p)
        if s is None:
            raise Exception()
    except:
        test_error("Could not create newly registered scheduler type")


def test_wrapper_api():
    from sprokit.pipeline import config
    from sprokit.pipeline import modules
    from sprokit.pipeline import pipeline
    from sprokit.pipeline import process_registry
    from sprokit.pipeline import scheduler_registry

    sched_type = 'python_example'
    sched_desc = 'simple description'

    modules.load_known_modules()

    reg = scheduler_registry.SchedulerRegistry.self()

    reg.register_scheduler(sched_type, sched_desc, example_scheduler(False))

    p = pipeline.Pipeline()

    preg = process_registry.ProcessRegistry.self()

    proc_type = 'orphan'
    proc_name = 'orphan'

    proc = preg.create_process(proc_type, proc_name)

    p.add_process(proc)

    def check_scheduler(s):
        if s is None:
            test_error("Got a 'None' scheduler")
            return

        s.start()
        s.pause()
        s.resume()
        s.stop()
        s.start()
        s.wait()

        del s

    p.reset()
    p.setup_pipeline()

    s = reg.create_scheduler(sched_type, p)
    check_scheduler(s)


if __name__ == '__main__':
    import os
    import sys

    if not len(sys.argv) == 4:
        test_error("Expected three arguments")
        sys.exit(1)

    testname = sys.argv[1]

    os.chdir(sys.argv[2])

    sys.path.append(sys.argv[3])

    tests = \
        { 'import': test_import
        , 'create': test_create
        , 'api_calls': test_api_calls
        , 'register': test_register
        , 'wrapper_api': test_wrapper_api
        }

    from sprokit.test.test import *

    run_test(testname, tests)
